### PR TITLE
support loading COG layers from query params

### LINKED ISSUE #9527: support loading COG layers from query params
## Description

now that we have COG support since #9320 im looking at how to implement support for loading such data type from the url params eg http://localhost:8081/#/viewer/new?actions=[{%22type%22:%22CATALOG:ADD_LAYERS_FROM_CATALOGS%22,%22layers%22:[%22title%22],%22sources%22:[{%22type%22:%22cog%22,%22url%22:%22https%3A%2F%2Fcogeo.craig.fr%2Fopendata%2Fortho%2Forthocraig3_vichy_2021.cog.tif%22}]}]

for now i only changed one thing so that the layer title is not looked for in the catalog records like its done for 3dtiles (#9021) & geojson (#9276):
```
--- a/web/client/epics/catalog.js
+++ b/web/client/epics/catalog.js
@@ -163,7 +163,7 @@ export default (API) => ({
                         const layerOptionsParam = get(options, i, searchOptionsSelector(state));
                         // use the selected layer text as title for 3d tiles
                         // because currently we get only a single record for this service type
-                        const layerOptions = format === '3dtiles' || format === 'geojson'
+                        const layerOptions = format === '3dtiles' || format === 'geojson' || format === 'cog'
                             ? {
                                 ...layerOptionsParam,
                                 title: isObject(layerOptionsParam?.title)

```

but no layer is loaded. From my digging in the browser debugger, i understand that's because `service.records` is empty in https://github.com/geosolutions-it/MapStore2/blob/master/web/client/api/catalog/COG.js#L65, probably because `searchAndPaginate` hasnt been called yet, and that method will return no records anyway since `layers` is an empty array ? so ... does it make sense to have a `searchAndPaginate` method in a COG catalog type ? afaict this method isnt present for `geojson` and `3dtiles` formats... @dsuren1, any idea ?

### PR DESCRIPTION
**Pull request by landryb**
## Description
implements #9527

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**

fixes #9527

**What is the new behavior?**
http://localhost:8081/#/viewer/new?actions=%5B%7B%22type%22:%22CATALOG:ADD_LAYERS_FROM_CATALOGS%22,%22layers%22:%5B%22testtitle%22%5D,%22sources%22:%5B%7B%22type%22:%22cog%22,%22url%22:%22https%3A%2F%2Fcogeo.craig.fr%2Fopendata%2Fortho%2Forthocraig3_vichy_2021.cog.tif%22%7D%5D%7D%5D should work and zoom to the provided COG

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Comment by landryb** [comment]
@dsuren1 help/feedback welcome on this. Even with `projectionDefs` containing `EPSG:2154` i cant get the layer to be automagically added.

**Comment by landryb** [comment]
ok it loads the cog with your tweak from https://github.com/geosolutions-it/MapStore2/issues/9527#issuecomment-1742711134 but im not comfy with it.. maybe i should set name to the value of `text` to avoid having a specialcase like this or that can lead to conflicts later on ?

also, it doesn't automatically zoom to the COG extent even if it knows its bbox (i have the 'zoom to layer' button in the toc), while it's the case when the COG is added from the catalog.

**Comment by landryb** [comment]
with 9f2f3c93a it works but i have to figure out what triggers the zoom to layer behaviour

**Comment by landryb** [comment]
> Thank you so much for contributing @landryb. We will check this asap. This is still WIP n your side as reported din description?

no this is ready for review, eslint is happy with the js changes, and i've just tested it backported on a local 2023.02 build of ms2-georchestra and it does what i expect, eg `?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["cog clerco cc46"],"sources":[{"type":"cog","url":"https%3A%2F%2Fcogeo.craig.fr%2Fopendata%2Fortho%2Fortho25_cc46_clerco_1996.cog.tif"}]}]` properly loads the COG at initial ms2 loading.

**Comment by landryb** [comment]
updated the issue description to state that it wasnt just WIP, but working for me :) it was already working in the initial PR from october, but it had stalled and bitrotted.

**Comment by landryb** [comment]
can i get a review on this before it bitrots more ? should i rebase the branch or a merge commit like [f22d540](https://github.com/geosolutions-it/MapStore2/pull/9531/commits/f22d5405b1745d4c221ded6c6355ac5cc9137779) above is acceptable ?

**Comment by dsuren1** [comment]
@landryb we are waiting for this [PR](https://github.com/geosolutions-it/MapStore2/pull/9857) to be merged as it refactors the COG workflow along with options to style. I think you may have to rebase again to include the new changes after the linked PR is merged.

**Comment by dsuren1** [comment]
@landryb Could you please align your changes with latest master. Thanks!

**Comment by landryb** [comment]
> @landryb Could you please align your changes with latest master. Thanks!

sure, will try to understand what happened in #9857 & #10089 wrt my changes.. from my understanding i should reuse `LayerUtils.getLayerConfig` instead of my own `fetchMetadata` (added in 461da074, used in 71899b36c) .. and test.

**Comment by landryb** [comment]
@dsuren isnt there a problem with `LayerUtils.getLayerConfig` added in d683b824 ? the order of parameters doesn't seem to be match between the call in https://github.com/geosolutions-it/MapStore2/blob/master/web/client/api/catalog/COG.js#L57:
```
return LayerUtils.getLayerConfig({url, controller, layer})
```
and the API in https://github.com/geosolutions-it/MapStore2/blob/master/web/client/utils/cog/LayerUtils.js#L56
```
export const getLayerConfig = ({ url, layer, controller })
```

i dunno if that matters.. but whatever order of parameters i put in my code, eslint complains if i try building the layer object/parameter on the fly:
```
Line 72, column 52: /data/src/georchestra/mapstore2-georchestra/MapStore2/web/client/api/catalog/COG.js: Unexpected token (72:52)

  70 |     return Promise.all([...layers]).then((_layers) => {
  71 |         if (!_layers.length) {
> 72 |             return LayerUtils.getLayerConfig({_url, {
```
eslint is happy if i build the layer object before, as done in [0260ac7](https://github.com/geosolutions-it/MapStore2/pull/9531/commits/0260ac7eaadceb0de0de64a10c5bc7d371c5c8c4) (untested yet)

**Comment by dsuren1** [comment]
@landryb 

`getLayerConfig` takes object as a parameter, the order of the prop is not a concern but the name of the prop passed. Modify your code as shown below

```diff
> 80 |        return LayerUtils.getLayerConfig({url: _url, layer, controller})
```

**Comment by landryb** [comment]
> @landryb
> 
> `getLayerConfig` takes object as a parameter, the order of the prop is not a concern but the name of the prop passed. Modify your code as shown below
> 
> ```diff
> return LayerUtils.getLayerConfig({url: _url, layer, controller})
> ```

ah, thanks, now i understand the logic, it wasnt obvious for me that the variable names were auto matched on object properties. fixed with [60e7269](https://github.com/geosolutions-it/MapStore2/pull/9531/commits/60e72699e3245bfad56043ad57e52225efc0686b) and i've been able to successfully test it on a locally running instance with `?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["Vichy 3cm"],"sources":[{"type":"cog","url":"https%3A%2F%2Fcogeo.craig.fr%2Fopendata%2Fortho%2Forthocraig3_vichy_2021.cog.tif"}]}]` as params

**Review by dsuren1 — CHANGES_REQUESTED** [review]
@landryb 
Kindly update the doc`map-query-parameters.md` for adding COG layer via query paramater similar to other layer type. Also a note in the docs regarding the 'limitation' mentioned below. Thanks!


@offtherailz @tdipisa 
The only limitation to adding COG layer via query parameter, is that it is not possible to cancel the metadata fetch call as there is not control exposed at the moment. So when a big source is requested the map load time maybe significantly affected.

File: docs/developer-guide/map-query-parameters.md [hunk 1]
```diff
@@ -372,6 +372,20 @@ GET `#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers
...
+{
+    "type": "CATALOG:ADD_LAYERS_FROM_CATALOGS",
+    "layers": ["My huge remote satellite COG"],
+    "sources": [{ "type":"cog", "url":"https://example.com/satellite_imagery.tif" }]
+}
+```
+
+GET: `#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["My huge remote satellite COG"],"sources":[{"type":"cog","url":"https://example.com/satellite_imagery.tif"}]}]`
+
+Depending on the internal structure optimization done on the remote COG file, the map load time might be long.
```
Comment (dsuren1) [inline docs/developer-guide/map-query-parameters.md:387]:
```suggestion
!!! note
    Depending on the internal structure optimization done on the remote COG source, the map load time might be long. Furthermore, it is not feasible to cancel metadata fetching for the COG layer(s) when loading layers via query parameters.
```

**Review by dsuren1 — CHANGES_REQUESTED** [review]

**Review by dsuren1 — APPROVED** [review]

**Comment by dsuren1** [comment]
@ElenaGallo Kindly test it in DEV and let us know if it can be backported. Thanks!

**Comment by ElenaGallo** [comment]
test passed, @dsuren1 please backport to 2024.01.xx. Thanks

**Comment by tdipisa** [comment]
@ElenaGallo @dsuren1 this is for 2024.01.01 so let's wait the delivery of 2024.01.00 before merging the backport.

**Pull request merged**